File: Android/app/src/main/java/ru/inwords/inwords/data/source/webService/WebRequestsManager.java
package ru.inwords.inwords.data.source.webService;

import java.util.List;

import io.reactivex.Single;
import ru.inwords.inwords.data.dto.EntityIdentificator;
import ru.inwords.inwords.data.dto.User;
import ru.inwords.inwords.data.dto.UserCredentials;
import ru.inwords.inwords.data.dto.WordTranslation;
import ru.inwords.inwords.data.dto.game.Game;
import ru.inwords.inwords.data.dto.game.GameInfo;
import ru.inwords.inwords.data.dto.game.GameLevel;
import ru.inwords.inwords.data.dto.game.LevelScore;
import ru.inwords.inwords.data.dto.game.LevelScoreRequest;
import ru.inwords.inwords.data.source.webService.session.TokenResponse;
import ru.inwords.inwords.data.sync.PullWordsAnswer;

public interface WebRequestsManager {
    Single<TokenResponse> getToken(UserCredentials userCredentials);

    Single<TokenResponse> registerUser(UserCredentials userCredentials);

    Single<String> getLogin();

    Single<User> getAuthorisedUser();

    Single<User> getUserById(int id);

    Single<List<EntityIdentificator>> insertAllWords(List<WordTranslation> wordTranslations);

    Single<Integer> removeAllServerIds(List<Integer> serverIds);

    Single<PullWordsAnswer> pullWords(List<Integer> serverIds);

    Single<List<GameInfo>> getGameInfos();

    Single<Game> getGame(int gameId);

    Single<GameLevel> getLevel(int levelId);

    Single<LevelScore> getScore(LevelScoreRequest levelScoreRequest);

    Single<Boolean> uploadScore(List<LevelScoreRequest> levelScoreRequests);
}

File: Android/app/src/main/java/ru/inwords/inwords/data/source/webService/WebRequestsManager.kt
package ru.inwords.inwords.data.source.webService

import io.reactivex.Single
import ru.inwords.inwords.data.dto.EntityIdentificator
import ru.inwords.inwords.data.dto.User
import ru.inwords.inwords.data.dto.UserCredentials
import ru.inwords.inwords.data.dto.WordTranslation
import ru.inwords.inwords.data.dto.game.*
import ru.inwords.inwords.data.dto.google.TtsSynthesizeRequest
import ru.inwords.inwords.data.source.webService.session.TokenResponse
import ru.inwords.inwords.data.sync.PullWordsAnswer

interface WebRequestsManager {
    fun getLogin(): Single<String>

    fun getAuthorisedUser(): Single<User>

    fun getGameInfos(): Single<List<GameInfo>>

    fun getUserEmail(): Single<String>

    fun getToken(): Single<TokenResponse>

    fun getToken(userCredentials: UserCredentials): Single<TokenResponse>

    fun registerUser(userCredentials: UserCredentials): Single<TokenResponse>

    fun getUserById(id: Int): Single<User>

    fun insertAllWords(wordTranslations: List<WordTranslation>): Single<List<EntityIdentificator>>

    fun removeAllServerIds(serverIds: List<Int>): Single<Int>

    fun pullWords(serverIds: List<Int>): Single<PullWordsAnswer>

    fun getGame(gameId: Int): Single<Game>

    fun getLevel(levelId: Int): Single<GameLevel>

    fun getScore(levelScoreRequest: LevelScoreRequest): Single<LevelScore>

    fun uploadScore(levelScoreRequests: List<LevelScoreRequest>): Single<Boolean>

    fun ttsSynthesize(request: TtsSynthesizeRequest, googleServicesApiKey: String): Single<String>
}
